iceoryx_utils/include/iceoryx_utils/internal/units/duration.inl

```diff
@@ -20,262 +20,511 @@ namespace iox
+
+    uint64_t multiplicatorHigh = static_cast<uint32_t>(multiplicator >> 32U);
+    auto nanosecondsFromHigh = m_nanoseconds * multiplicatorHigh;
+    auto fullBlocksOfSecondsOnly = nanosecondsFromHigh / ONE_FULL_BLOCK_OF_SECONDS_ONLY;
+    auto remainingBlockWithFullAndFractionalSeconds = nanosecondsFromHigh % ONE_FULL_BLOCK_OF_SECONDS_ONLY;
+
+    auto durationFromNanosecondsHigh = Duration{fullBlocksOfSecondsOnly * SECONDS_PER_FULL_BLOCK, 0U}
+                                       + Duration::nanoseconds(remainingBlockWithFullAndFractionalSeconds << 32);
+
+    return durationFromSeconds + durationFromNanosecondsLow + durationFromNanosecondsHigh;
 }
 
-inline constexpr Duration operator"" _us(unsigned long long int value) // PRQA S 48
+
+template <typename From, typename To>
+inline constexpr bool Duration::wouldCastFromFloatingPointProbablyOverflow(const From floatingPoint) const noexcept
 {
-    return Duration{static_cast<long double>(value) / 1000000.0};
+    static_assert(std::is_floating_point<From>::value, "only floating point is allowed");
+
+    // depending on the internal representation this could be either the last value to not cause an overflow
+    // or the first one which causes an overflow;
+    // to be safe, this is handled like causing an overflow which would result in undefined behavior when casting to
+    // Seconds_t
+    constexpr To SECONDS_BEFORE_LIKELY_OVERFLOW = static_cast<To>(std::numeric_limits<Seconds_t>::max());
+    return floatingPoint >= SECONDS_BEFORE_LIKELY_OVERFLOW;
 }
 
-inline constexpr Duration operator"" _ms(long double value)
+template <typename T>
+inline constexpr Duration Duration::fromFloatingPointSeconds(const T floatingPointSeconds) const noexcept
 {
-    return Duration{value / 1000.0};
+    static_assert(std::is_floating_point<T>::value, "only floating point is allowed");
+
+    if (std::isinf(floatingPointSeconds))
+    {
+        std::clog << __PRETTY_FUNCTION__ << ": Multiplication with Inf, clamping to max value!" << std::endl;
+        return Duration::max();
+    }
+
+    double secondsFull{0.0};
+    double secondsFraction = modf(floatingPointSeconds, &secondsFull);
+
+    if (wouldCastFromFloatingPointProbablyOverflow<T, Seconds_t>(secondsFull))
+    {
+        std::clog << __PRETTY_FUNCTION__ << ": Result of multiplication would overflow, clamping to max value!"
+                  << std::endl;
+        return Duration::max();
+    }
+
+    return Duration{static_cast<Seconds_t>(secondsFull),
+                    static_cast<Nanoseconds_t>(secondsFraction * NANOSECS_PER_SEC)};
 }
 
-inline constexpr Duration operator"" _ms(unsigned long long int value) // PRQA S 48
+template <typename T>
+inline constexpr Duration
+Duration::multiplyWith(const std::enable_if_t<std::is_floating_point<T>::value, T>& rhs) const noexcept
 {
-    return Duration{static_cast<long double>(value) / 1000.0};
+    // operator*(...) takes care of negative values for rhs
+
+    if (std::isnan(rhs))
```

Comment:
When you assume that not a number is like infinity, what about the case when Duration == 0? In this case duration should not be set to max - after the operation it should still be zero.

Reply:
done